### ISSUE TITLE
[CLOSED] Operator Refactoring/Split-Up Proposal

### ISSUE THREAD

**Issue by msohn** [msohn/issuemigration#79]
<a href="https://github.wdf.sap.corp/i068969"><img src="https://github.wdf.sap.corp/avatars/u/4644?" align="left" width="96" height="96" hspace="10"></img></a> **Issue by [i068969](https://github.wdf.sap.corp/i068969)**
_Thursday Aug 17, 2017 at 14:50 GMT_
_Originally opened as https://github.wdf.sap.corp/kubernetes-attic/garden-operator/issues/79_

----

# Refactoring proposal v2

tl;dr garden-operator is too complex; split it in multiple controllers in Seed clusters that manage `Helm Releases` and `Terraform` Jobs.

## Current issues

In the current state garden-operator is a complex monolith beast. Its architecture won't scale with hundreds / thousands of clusters and many dozens seed clusters. Checking if the cluster is in correct state can involve lots of conditional logic. Putting all the logic in one place can be a short win, but in the long run it will be a lost cause as complexity increases.

Installation of infrastructure / addons / control plane happens via go-lang templates, Helm charts and Terraform jobs with go-lang templates. Cleanup operations for said components can only be achieve by writing more conditional logic in the garden-operator.

### Garden-operator

Garden-operator should act as a federation controller - only do CRUD operations on TPR/CRD(s) in Seed clusters. This will greatly reduce network latency to the API servers and CloudProvider's APIs and the occasional SAP network outages / hiccups. All the logic for CRUD of Shoot clusters is dedicated to Seed clusters.

### New TPR/CRD & Operators

Several TPR/CRD should be created in each Seed and Shoot cluster. All controllers bellow should try to apply the [best practices for writing controllers](https://github.com/kubernetes/community/blob/master/contributors/devel/controllers.md):

#### HelmRelease TPR/CRD

`HelmRelease` is a Helm Chart release installed and managed in a cluster (any cluster):

```
apiVersion: garden.k8s.sapcloud.io/v1
kind: HelmRelease
metadata:
  name: control-plane-stack
  namespace: my-shoot-cluster
spec:
  chart:
    path: stable/control-plane-stack | /some-dir/control-plane-stack | https://chart.repo/control-plane-1.2.3.tgz
    version: 1.2.3
    # secret containing the custom values for the chart
    # optional
    valuesSecret: control-plane-stack-secret
status:
  # every status that is defined in
  # https://github.com/kubernetes/helm/blob/4c1a47480e70cabb5d9e227bc55bd75ea9406de0/pkg/proto/hapi/release/status.pb.go#L19-L36
  helmStatus: Unkown | Deployed | Deleted | Suspended | Failed | Deleting | PendingInstall | PendingUpgrade | PendingRollback
  lastWorkingRevision: 4
```

#### HelmReleaseOprator

`HelmReleaseOperator` handles create/update/delete operations on the Tiller server.

Two K8S API resources are monitored for changes (using informers and stores):
- `HelmRelease`
- `Secrets` (any modification of a secret required in a `HelmRelease` must trigger upgrade)

Scenario:

> Note: All scenarios bellow happen in a seed cluster

Create `HelmRelease`:
- `kubectl create -f control-plane-stack.yaml`
- operator sees that a new `HelmRelease` resource is created, finds the chart (mounted via ConfigMap / downloaded from some Helm Repo), finds the secret associated with that release and calls the Tiller server to install the needed chart. It sets the Status in `helmStatus`.

Update `HelmRelease`:
- `kubectl apply -f control-plane-stack-v2.yaml`
- operator sees that an existing `HelmRelease` is modified checks the status of the running `HelmRelease` and does a helm Upgrade (or install if the release is deleted for some reason).

Delete `HelmRelease`:
- `kubectl delete -f control-plane-stack-v2.yaml`
- operator sees that Helm Release wants to be deleted and calls Tiller for deletion with purge.

Additionally there is a `resync` loop that periodically calls Tiller and updates the `helmStatus` of the releases (resources may be deleted / not created) triggering install / upgrade / rollback of a release. This loop may run in different binary or be part of the `HelmReleaseOperator`. It's only job is to update the status of the corresponding HelmRelease. Using this loop we can guarantee that resources are always available for a `HelmRelease`.

#### Terraform TPR/CRD

`Terraform` TPR/CRD handles changes to infrastructure:

```
apiVersion: garden.k8s.sapcloud.io/v1
kind: Terraform
metadata:
  name: worker-infra
  namespace: my-shoot-cluster
spec:
  chart:
    # secrets for the terraform - AWS keys etc. (optional)
    secret: worker-infra-secret
    # Terraform workload
    config: worker-infra-config
status: Applied | Applying | Destroying | NeedsApply | Failed
```

#### TerraformOperator

`TerraformOperator` handles create/update/delete operations based on `Terraform` configuration:

Four K8S API resources are monitored for changes (using informers and stores):
- `Terraform`
- `Job` (and CronJob when it hits [beta in 1.8](https://github.com/kubernetes/kubernetes/issues/41039#issuecomment-320960821))
- `Secret` (any modification of a secret required in a `Terraform` job must trigger `terraform apply`)
- `ConfigMap` (any modification of a configmap required in a `Terraform` job must trigger `terraform apply`)

Scenario:

> Note: All scenarios bellow happen in a seed cluster

Create `Terraform`:
- `kubectl create -f worker-infra.yaml`
- operator sees that a new `Terraform` resource is created. It creates a Job with all the needed configurations and secrets. The Job executes `terraform apply ....`(via `run.sh`). Once the job is finished record the status in `status`.

Update `Terraform`:
- `kubectl apply -f worker-infra.yaml`
- operator sees that a existing `Terraform` resource is updated. It executes the same steps as in Create `Terraform` step

Delete `Terraform`:
- `kubectl delete -f worker-infra.yaml`
- operator sees that a existing `Terraform` resource is deleted. It creates a Job with all the needed configurations and secrets. The Job executes `terraform destroy ....`(via `run.sh`). Once the job is finished record the status in `status`.

Additionally there is a `resync` loop that periodically changes the status of the `Terraform` resource to `NeedsApply`, triggering the operator to perform `terraform apply...`. This can be achieved via CronJob in the 1.8

#### DNSTerraformOperator

This is similar to [ExternalDNS](https://github.com/kubernetes-incubator/external-dns), but we can't use it due to the lack of support for OpenStack(is this a requirement?)

Three K8S API resources are monitored for changes (using informers and stores):
- `Terraform`
- `Service`
- `ConfigMap`

Once a service is annotated with `external-dns.alpha.garden.k8s.sapcloud.io`
```
kubectl expose deployment kube-apiserver --port=443 --target-port=443 --type=LoadBalancer
kubectl annotate service kube-apiserver "external-dns.alpha.garden.k8s.sapcloud.io/hostname=api.my-cluster.k8s.sapcloud.io."
```

The operator creates a `Terraform` resource with the ELB's endpoint as a configuration. The `Terraform` job automatically creates the needed `Route53` resource and starts to reconcile.

### Operators cont

All operators can be bundled together in a single binary (like k8s's `controller-manager`) or deployed independently from one another.

## Cluster Create Scenario

* A new cluster is created in the Garden-UI
* Shoot resource is created
* Garden-Operator selects a Seed Cluster and creates all needed secrets, HelmRelease....
* as a prerequisite in Seed both `Tiller` and previous mentioned operators are deployed
* Multiple `HelmRelease`s with control plane components, monitoring stack, [addon-manager](https://github.com/kubernetes/kubernetes/tree/master/cluster/addons/addon-manager)...  are deployed
  > Note: addon-manager will connect to its own k8s Shoot api server
* In the control plane `HelmRelease` there is a nested `Terraform` that creates the infra for the shoot cluster.
  > Note this nested `Terraform` will play a role in the Cluster deletion operation later on.
* Shoot's API server is annotated with `external-dns.alpha.garden.k8s.sapcloud.io/hostname=api.my-cluster.k8s.sapcloud.io.`
* in addon-manager's mounted ConfigMaps are all managed by us resources that needs to be created and Reconciled in the Shoot cluster, including:
  * Tiller
  * mentioned before operators
  * `HelmReleases` with different operators and optional `Terraform` resources

### Prerequisites

The delete cluster scenario is special because we need to do several things, before we tear down the infra we create for workers:

* both `Tiller` and previous mentioned operators needs to be deployed
* Stop `addon-manager`
* Delete all `HelmReleases` in the Shoot cluster - helm charts can have `pre-delete` and `post-delete` hooks to clean up resources created in some third-party systems (think accounts in SuccessFactors, Ariba...., credentials ....)
* Delete all `Terraform` resources to clean up CloudProvider resources created by addons in the shoot cluster
  > Note: previous 2 steps can be a `post-delete` step in the deletion of `addon-manager`

### Scenario:

* An exising cluster is deleted in the Garden-UI
* Shoot resource is scheduled for deletion
* See Prerequisites
* Delete the Shoot Cluster namespace, triggering deletion of all `HelmReleases` and `Terraform`

## Cluster Update Scenario

The only tricky part for the Upgrade scenario is `kubelet` upgrades. In our current setup this can only be achieved by modifying the launch configuration for nodes. The rest of the components can be upgraded by simply modifying all `HelmReleases`, `ConfigMaps`, `Secrets` and `Terraform`.

This will also allow us to do a rolling upgrade of Shoot-clusters from the garden operator.

Edits:
[0] Fixed typos

**Comment by msohn** [comment]
<a href="https://github.wdf.sap.corp/d043832"><img src="https://github.wdf.sap.corp/avatars/u/4487?" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [d043832](https://github.wdf.sap.corp/d043832)**
_Saturday Aug 19, 2017 at 10:52 GMT_

----

> In the current state garden-operator is a complex monolith beast.
 
I find that a bit harsh. See also Fowler on this: https://martinfowler.com/bliki/MonolithFirst.html
 
- Almost all the successful microservice stories have started with a monolith that got too big and was broken up
- Almost all the cases where I've heard of a system that was built as a microservice system from scratch, it has ended up in serious trouble.
 
See, I am not favouring a "complex monolith beast", but I agree with Fowler. First, let's understand what we need to do and where the demarcations are, then we can think about splitting it up.
 
Also, for my money, we are not in the position to do that now, because on the one hand, we have a very challenging beast ahead of us, that will surely bring huge disruptions to the operator: https://github.wdf.sap.corp/kubernetes/garden-operator/issues/53 -> https://github.wdf.sap.corp/kubernetes/garden-operator/issues/56 (you need to be authenticated to see the images which are still in Jira) -> https://github.wdf.sap.corp/kubernetes/garden-operator/issues/27. That will be the next big milestone after the self-service. On the other hand, we lack practical stuff, like automated build/CI.
 
Right now, I am not willing to split up the operator. We neither understand the target picture very well nor do we have the infrastructure to handle a multitude of controllers in a multitude of clusters. It's simply too early for that.
 
> Its architecture won't scale with hundreds / thousands of clusters and many dozens seed clusters.
 
Don't get me wrong, but where is that insight coming from? See, a.) the operator isn't doing much itself (it merely only "controls" stuff, but the heavy lifting is done elsewhere) and b.) it can/should be scalable. We need to handle multiple operator instances well. See also thoughts on failing/multiple operator instances in https://github.wdf.sap.corp/kubernetes/garden-operator/issues/25. That will buy us even more time (scale out) and needs to be done anyway (rolling update, failure and restart). That's for me far more important.
 
> Putting all the logic in one place can be a short win, but in the long run it will be a lost cause as complexity increases.
 
That's the simplest thing we can factor out, but later, please. We need to wait until we have the proper infrastructure ready to support us painlessly.
 
> Garden-operator should act as a federation controller
 
Look, it's easily said, but there is huge effort in writing a "federation" or what I would rather call a "coordination" controller. Loosely-coupled systems are OK, but as you said yourself, that's not a loosely coupled collection of controllers, but there is the need to coordinate the work of multiple controllers (and order between the steps, so decoupling this is hard). The monolith that you don't like so much, is not that bad in this phase of the project.
 
This will greatly reduce network latency to the API servers and CloudProvider's APIs and the occasional SAP network outages / hiccups.
 
Again, where did you see that we have any issue at all with network latency? This smells to me like "premature performance optimization". The network outages are a real thing, but we have outages also on AWS. We need to be resilient anyway.
 
When we see, we have too many problems with the Cluster IV OpenStack later, we must analyse whether a decentralized approach will help us at all (which I don't see), because as you said, even if you split up the operator, the coordinator and the operations UI runs on this OpenStack infrastructure nevertheless. If these parts aren't working, it's of little help if other parts work. The principle problem is to deploy to our infrastructure in the first place.

> Tiller
 
I see your push for Helm/Tiller and I can understand that in parts (the chart approach itself), but honestly, Tiller is a terrible piece of ugly technology violating Kubernetes core principles. This thing is a pain in the ass for me, it fails regularly, it's horribly implemented. I mean really, port-forwarding from the client through the API server to Tiller? WTF, why isn't that a clean controller acting on chart and release CRDs? The answer is, I believe, only historic. Would someone rewrite it today, Tiller would be implemented differently and more reliable.
 
Having said that, if you guys want to go for that, it's OK (it's not a redesign like the controller split-up, but an implementation detail). We use it already, so it doesn't matter if we do more of it. But I don't trust it much. Hopefully someone will pick up the task and keep the charts, but modernize the broken design and implementation of Tiller one day. If you go there more, we should understand it better. E.g., I am not sure what Tiller does when e.g. configmaps and the like were manipulated (unexpectedly) and whether they are updated. In any case, if we go more and more into Tiller, let's better understand what it does, when and how.
 
> Multiple controllers/ HelmReleases …
 
Well, there are dependencies. We need to do things first, before we can take the result and do other things and so it goes on. You can't run that in parallel. That's not possible because of the interdependencies, so if you do it, you must implement loose coupling, which is effort. Again, we won't refactor the operator into multiple pieces as someone needs then to take care of that. That's a complex task. It isn't easy to write micro services and a loosely coupled system.

> Cluster Update Scenario, The only tricky part for the Upgrade scenario is kubelet upgrades.
I think that's a total underestimation of the problems we will face with updates. The charts are not the problem.
 
We will keep your story in the backlog and we might work on that, but not now. Maybe in 6 or 12 months from now, maybe earlier when there is need to, but I am not willing to go there now. It makes everything much more complex for the above reasons:
1. Our priorities are the self-service, then automated updates - the refactoring of the operator doesn't help with either and makes the latter even harder, before we even know how to do that.
2. We lack infrastructure for the split-up.
3. The monolith is easier to build and maintain, because of the inter-dependencies between the different processing steps. For now at least/still.
4. Only when we have solved the hard problems and understand our monolith well enough, we can refactor it.
5. Your main argument is performance which I am not accepting at all right now - what I would accept is extensibility, maintainability, resilience, reliability and robustness. While I see that a split-up helps with extensibility, I see the split-up fail on the other dimensions. With no infrastructure it's hard to maintain. Decoupling a strictly interdependent system into a loosely coupled system is hard to grasp and maintain (one page of flow logic code in the operator) and the 3 R's are harder to implement.
 
I hope you are not too disappointed. Let's rather finish what we have started, before we engage in a huge refactoring right now. I am not piling up technical debt here - I think It's reasonable to wait with such a thing until we have the infrastructure and understanding to do it well.

**Comment by msohn** [comment]
<a href="https://github.wdf.sap.corp/d043832"><img src="https://github.wdf.sap.corp/avatars/u/4487?" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [d043832](https://github.wdf.sap.corp/d043832)**
_Wednesday Nov 08, 2017 at 04:40 GMT_

----

If we are to open source the Gardener and expect to see contributions we should make this an easy process for the contributors and us. While this proposal covers the "technical" split-up I am more concerned about the "logical" split-up. Right now, all components that constitute the operator must be in code in our source repo. That does not fit a pluggable architecture. So, if for instance someone likes to add DigitalOcean support, it should not be necessary that it gets added to the operator repo, but it should be a separate entity deployed into the vicinity of the operator and configured appropriately. Consequently, the botanist abstraction should be turned into a sub operator/controller entity with a dedicated contract designed for extensibility. The community may see other areas of cooperation, collaboration and contribution.

In other words, before we engage in an expensive refactoring we should better understand the target goal of what we must do to not only better manage the operator in terms of size and complexity, but also what it takes to invite and enable the community in jointly growing the Gardener as a whole.

**Comment by msohn** [comment]
<a href="https://github.wdf.sap.corp/d043832"><img src="https://github.wdf.sap.corp/avatars/u/4487?" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [d043832](https://github.wdf.sap.corp/d043832)**
_Tuesday Dec 19, 2017 at 11:54 GMT_

----

I guess by now this ticket is no longer up-to-date. We had some refactorings, automated update is still not implemented and by the time it will the operator will have changed even more, we are considering API server extensions, will have to introduce new abstractions with the Open Sourcing, might kick out Terraform or don't want to invest into refactoing of the same, etc., so closing this ticket now.
